File: prototype/rails/src/test/delivery/magic_link_delivery_test.rb
require "test_helper"

class MagicLinkDeliveryTest < ActiveSupport::TestCase
  test "the prototype default flashes the link" do
    assert_instance_of FlashMagicLinkDelivery, build_with("flash")
  end

  test "mail is selected by name" do
    assert_instance_of MailMagicLinkDelivery, build_with("mail")
  end

  test "an unknown setting is refused rather than silently dropping links" do
    assert_raises(ArgumentError) { build_with("carrier_pigeon") }
  end

  private

  def build_with(setting)
    original = Rails.configuration.x.magic_links.delivery
    Rails.configuration.x.magic_links.delivery = setting

    MagicLinkDelivery.build(ActionDispatch::Flash::FlashHash.new)
  ensure
    Rails.configuration.x.magic_links.delivery = original
  end
end

File: prototype/node/src/app/delivery/magic-link-delivery.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { flashMagicLinkDelivery } from './flash-magic-link-delivery.ts'
import { mailMagicLinkDelivery } from './mail-magic-link-delivery.ts'
import { MAGIC_LINK_DELIVERIES, selectMagicLinkDelivery } from './magic-link-delivery.ts'

test('the prototype default flashes the link', () => {
  assert.equal(selectMagicLinkDelivery('flash'), flashMagicLinkDelivery)
})

test('mail is selected by name', () => {
  assert.equal(selectMagicLinkDelivery('mail'), mailMagicLinkDelivery)
})

test('every configurable name selects a delivery', () => {
  for (const name of MAGIC_LINK_DELIVERIES) {
    assert.equal(typeof selectMagicLinkDelivery(name).deliver, 'function')
  }
})
